openmc/filter.py

```diff
@@ -0,0 +1,1810 @@
+        for subclass in cls._recursive_subclasses():
+            if group['type'][()].decode() == subclass.short_name.lower():
+                return subclass.from_hdf5(group, **kwargs)
+
+        raise ValueError("Unrecognized Filter class: '"
+                         + group['type'][()].decode() + "'")
+
+    @property
+    def bins(self):
+        return self._bins
+
+    @bins.setter
+    def bins(self, bins):
+        self.check_bins(bins)
+        self._bins = bins
+
+    @property
+    def num_bins(self):
+        return len(self.bins)
+
+    def check_bins(self, bins):
+        """Make sure given bins are valid for this filter.
+
+        Raises
+        ------
+        TypeError
+        ValueError
+
+        """
+
+        pass
+
+    def to_xml_element(self):
+        """Return XML Element representing the Filter.
+
+        Returns
+        -------
+        element : xml.etree.ElementTree.Element
+            XML element containing filter data
+
+        """
+        element = ET.Element('filter')
+        element.set('id', str(self.id))
+        element.set('type', self.short_name.lower())
+
+        subelement = ET.SubElement(element, 'bins')
+        subelement.text = ' '.join(str(b) for b in self.bins)
+
+        return element
+
+    def can_merge(self, other):
+        """Determine if filter can be merged with another.
+
+        Parameters
+        ----------
+        other : openmc.Filter
+            Filter to compare with
+
+        Returns
+        -------
+        bool
+            Whether the filter can be merged
+
+        """
+        return type(self) is type(other)
+
+    def merge(self, other):
+        """Merge this filter with another.
+
+        Parameters
+        ----------
+        other : openmc.Filter
+            Filter to merge with
+
+        Returns
+        -------
+        merged_filter : openmc.Filter
+            Filter resulting from the merge
+
+        """
+
+        if not self.can_merge(other):
+            msg = 'Unable to merge "{0}" with "{1}" '.format(
+                type(self), type(other))
+            raise ValueError(msg)
+
+        # Merge unique filter bins
+        merged_bins = np.concatenate((self.bins, other.bins))
+        merged_bins = np.unique(merged_bins)
+
+        # Create a new filter with these bins and a new auto-generated ID
+        return type(self)(merged_bins)
+
+    def is_subset(self, other):
+        """Determine if another filter is a subset of this filter.
+
+        If all of the bins in the other filter are included as bins in this
+        filter, then it is a subset of this filter.
+
+        Parameters
+        ----------
+        other : openmc.Filter
+            The filter to query as a subset of this filter
+
+        Returns
+        -------
+        bool
+            Whether or not the other filter is a subset of this filter
+
+        """
+
+        if type(self) is not type(other):
+            return False
+
+        for bin in other.bins:
```

Comment:
I would recommend renaming this to avoid confusion for readers with the Python built-in `bin()` function. Perhaps `_bin` or `other_bin`